### PR TITLE
*: fix update JSON field.

### PR DESCRIPTION
**Pull request by hicqu**
We have 2 `CAST STRING AS JSON`:  1 is use the string as JSON primitive (which is the default), 2 is parse the string as JSON. For `UPDATE table SET field = 's'`, If the field is `JSON`, we should `parse` it instead of use it as primitive.

### CODE REVIEW

File: plan/logical_plan_builder.go [hunk 1]
```diff
@@ -589,6 +589,7 @@ func (b *planBuilder) buildProjection4Union(u *LogicalUnionAll) {
 			srcType := srcCol.RetType
 			if !srcType.Equal(dstType) {
 				exprs[i] = expression.BuildCastFunction(b.ctx, srcCol.Clone(), dstType)
+				exprs[i].GetType().Flag |= mysql.ParseToJSONFlag
```
Comment (shenli) [inline plan/logical_plan_builder.go]: This is a bug fix?  Please add a test case for it.

Comment (hicqu) [reply plan/logical_plan_builder.go]: Added in `write_test.go`.

Comment (hicqu) [reply plan/logical_plan_builder.go]: For `union` it's hard to test because `dstType` == `JSON` will never happens.

**Comment by coocood** [comment]
@hicqu 
Is there a clear rule when we should add `ParseToJson` flag?
And how can we make sure all `ParseToJson` is properly set.

**Comment by hicqu** [comment]
@coocood you get the point. In fact I find that only one place we need to use string as JSON primitive value, which is in `json_set/json_insert/json_replace`. All of other places we just need to parse string into JSON. So I think change the default behavior to `parse` is good.

**Comment by coocood** [comment]
@hicqu 
So can we add the flag in `BuildCastFunct`?

**Comment by winoros** [comment]
The field type is stored by pointer and may point to the column's field type. So race will raise.

**Comment by hicqu** [comment]
@coocood Add a flag in `BuildCastFunc` or `WrapWithCastAsJson` is a good idea. But that needs change more code, and more time to do it. Could I do it later?

**Comment by hicqu** [comment]
@coocood @zhexuany PTAL again, thanks.

**Comment by coocood** [comment]
LGTM

@XuHuaiyu PTAL

**Pull request merged**